File: agent-service/src/agents/coordinator/scheduler.ts
import * as path from "path";
import simpleGit from "simple-git";
import { config } from "../../core/config";
import { decrypt } from "../../core/crypto";
import { redis } from "../../core/redis";
import { dataClient } from "../../services/dataClient";
import { OrchestrationService } from "../../services/orchestrationService";
import { AgentPool } from "./pool";
import type { AgentTask } from "./drivers/base";
import { sseManager } from "../../services/sseManager";

interface RepoConfig {
  id: string;
  name: string;
  remote_url: string;
  branch: string;
  auth_type: string;
  credentials: string;
  local_path?: string;
}

interface TaskConfig {
  id: string;
  title: string;
  description: string;
  type: string;
  status: string;
  agent_profile_id: string | null;
  depends_on_task_ids: string[];
  repository_ids: string[];
}

const TERMINAL_STATUSES = new Set(["done", "failed", "skipped", "cancelled"]);

export class Scheduler {
  private pool: AgentPool;

  constructor(pool: AgentPool) {
    this.pool = pool;
  }

  async run(projectId: string, planId: string): Promise<void> {
    // Mark the plan as executing — use generic patch so resume works regardless
    // of current plan status (avoids strict "confirmed" guard on startExecution).
    await dataClient.updatePlan(planId, { status: "executing" });
    this.emit(projectId, { type: "plan_executing", plan_id: planId });

    try {
      await this._runInternal(projectId, planId);

      await dataClient.completePlanExecution(planId);
      this.emit(projectId, { type: "plan_completed", plan_id: planId });
    } catch (err: unknown) {
      const reason = err instanceof Error ? err.message : String(err);
      try {
        await dataClient.failPlanExecution(planId, reason);
      } catch {
        // best-effort — don't shadow the original error
      }
      this.emit(projectId, { type: "plan_failed", plan_id: planId, error: reason });
      throw err;
    }
  }

  private async _runInternal(projectId: string, planId: string): Promise<void> {
    const repositoryList: RepoConfig[] = await dataClient.getProjectRepositories(projectId);
    const tasks: TaskConfig[] = await dataClient.getPlanTasks(planId);

    const workspaceMap = await this.cloneRepositories(projectId, repositoryList);
    const repositories = new Map(repositoryList.map((r) => [r.id, r]));

    // terminalIds — all terminal tasks (done/failed/skipped/cancelled).
    //   Used for loop-termination and filtering ready tasks.
    // completedIds — only "done" tasks.
    //   Used to satisfy dependency checks so failed tasks don't unblock dependents.
    const terminalIds = new Set<string>();
    const completedIds = new Set<string>();
    const inFlightIds = new Set<string>();

    // Pre-populate from DB state so resume works correctly.
    for (const t of tasks) {
      if (TERMINAL_STATUSES.has(t.status)) terminalIds.add(t.id);
      if (t.status === "done") completedIds.add(t.id);
    }

    while (terminalIds.size < tasks.length) {
      const ready = tasks.filter((t) => {
        if (terminalIds.has(t.id) || inFlightIds.has(t.id)) return false;
        return t.depends_on_task_ids.every((id) => completedIds.has(id));
      });

      if (ready.length === 0) {
        if (inFlightIds.size === 0) break; // deadlock or all tasks accounted for
        await new Promise((r) => setTimeout(r, 1000));
        continue;
      }

      const dispatches = ready.map((task) =>
        this.dispatchTask(
          projectId,
          planId,
          task,
          tasks,
          workspaceMap,
          repositories,
          terminalIds,
          completedIds,
          inFlightIds,
        )
      );

      await Promise.allSettled(dispatches);
    }
  }

  /**
   * Compute all transitive dependents of `taskId` within the loaded task list,
   * excluding tasks already in `excludeIds`.
   */
  private getTransitiveDependents(
    taskId: string,
    tasks: TaskConfig[],
    excludeIds: Set<string>,
  ): string[] {
    const result: string[] = [];
    const visited = new Set<string>();

    const traverse = (id: string): void => {
      const direct = tasks.filter(
        (t) => t.depends_on_task_ids.includes(id) && !excludeIds.has(t.id),
      );
      for (const dep of direct) {
        if (visited.has(dep.id)) continue;
        visited.add(dep.id);
        result.push(dep.id);
        traverse(dep.id);
      }
    };

    traverse(taskId);
    return result;
  }

  private async cloneRepositories(
    projectId: string,
    repositories: RepoConfig[]
  ): Promise<Map<string, string>> {
    const workspaceMap = new Map<string, string>();

    for (const repo of repositories) {
      const localPath = path.join(config.WORKSPACES_ROOT, projectId, repo.name);
      const cloneUrl = this.buildCloneUrl(repo);

      this.emit(projectId, {
        type: "repo_status",
        repo_id: repo.id,
        repo_name: repo.name,
        status: "cloning",
      });

      try {
        const git = simpleGit();
        try {
          await git.clone(cloneUrl, localPath, ["--branch", repo.branch ?? "main"]);
        } catch {
          await simpleGit(localPath).pull();
        }

        await dataClient.updateRepositoryStatus(repo.id, {
          status: "ready",
          local_path: localPath,
        });

        workspaceMap.set(repo.id, localPath);

        this.emit(projectId, {
          type: "repo_status",
          repo_id: repo.id,
          repo_name: repo.name,
          status: "ready",
        });
      } catch (err) {
        await dataClient.updateRepositoryStatus(repo.id, {
          status: "error",
          error_message: String(err),
        });
        this.emit(projectId, {
          type: "repo_status",
          repo_id: repo.id,
          repo_name: repo.name,
          status: "error",
          message: String(err),
        });
      }
    }

    return workspaceMap;
  }

  private buildCloneUrl(repo: RepoConfig): string {
    if (repo.auth_type === "token") {
      const token = decrypt(repo.credentials);
      const url = new URL(repo.remote_url);
      url.username = token;
      return url.toString();
    }
    return repo.remote_url;
  }

  /**
   * After a task completes, attempt to push any commits made in each workspace.
   * This is best-effort — failures are logged but do not fail the task.
   */
  private async pushWorkspaces(
    projectId: string,
    task: TaskConfig,
    workspaceMap: Map<string, string>,
    repositories: Map<string, RepoConfig>
  ): Promise<void> {
    for (const repoId of task.repository_ids) {
      const localPath = workspaceMap.get(repoId);
      const repo = repositories.get(repoId);
      if (!localPath || !repo) continue;

      try {
        const git = simpleGit(localPath);

        // Stage and commit any uncommitted changes the agent left behind
        const status = await git.status();
        if (!status.isClean()) {
          await git.add(".");
          await git.commit(`chore: agent result for task "${task.title}"`, { "--allow-empty": null });
        }

        // Push with the authenticated remote URL
        const pushUrl = this.buildCloneUrl(repo);
        await git.push(pushUrl, repo.branch ?? "main");

        this.emit(projectId, {
          type: "repo_status",
          repo_id: repo.id,
          repo_name: repo.name,
          status: "ready",
          message: "Changes pushed to remote",
        });
      } catch (err) {
        // Non-fatal — log and continue
        console.error(`[Scheduler] Failed to push repo ${repo.name}:`, err);
        this.emit(projectId, {
          type: "repo_status",
          repo_id: repo.id,
          repo_name: repo.name,
          status: "ready",
          message: `Push failed (non-fatal): ${String(err)}`,
        });
      }
    }
  }

  private async dispatchTask(
    projectId: string,
    planId: string,
    task: TaskConfig,
    allTasks: TaskConfig[],
    workspaceMap: Map<string, string>,
    repositories: Map<string, RepoConfig>,
    terminalIds: Set<string>,
    completedIds: Set<string>,
    inFlightIds: Set<string>
  ): Promise<void> {
    inFlightIds.add(task.id);

    await dataClient.updateTask(task.id, { status: "in_progress" });
    this.emit(projectId, {
      type: "task_status",
      task_id: task.id,
      status: "in_progress",
      agent_profile_id: task.agent_profile_id,
    });

    const workspaces: Record<string, string> = {};
    for (const repoId of task.repository_ids) {
      const localPath = workspaceMap.get(repoId);
      const repoName = repositories.get(repoId)?.name ?? repoId;
      if (localPath) workspaces[repoName] = localPath;
    }

    if (Object.keys(workspaces).length === 0) {
      const first = workspaceMap.values().next().value;
      if (first) workspaces["default"] = first;
    }

    const agentTask: AgentTask = {
      id: task.id,
      title: task.title,
      description: task.description,
      type: task.type,
      agentProfileId: task.agent_profile_id,
    };

    // ── Driver resolution: role-based first, then profile-based ──────────────
    const driver =
      this.pool.getDriverByRole(task.type) ??
      (task.agent_profile_id ? this.pool.getDriver(task.agent_profile_id) : null);

    this.emit(projectId, {
      type: "agent_started",
      task_id: task.id,
      agent_role: task.type,
      agent_profile_id: task.agent_profile_id ?? undefined,
    });

    let result;
    const startedAt = new Date().toISOString();

    if (driver) {
      result = await driver.execute(agentTask, workspaces);
    } else {
      result = { success: false, output: "", error: "No driver found for task type or profile" };
    }

    const completedAt = new Date().toISOString();

    // Capture git diff across all assigned workspaces
    const diffArtifacts: Array<{
      type: "diff";
      title: string;
      content: string;
      content_type: string;
      sort_order: number;
    }> = [];
    let diffIdx = 0;
    for (const repoId of task.repository_ids) {
      const localPath = workspaceMap.get(repoId);
      const repoName = repositories.get(repoId)?.name ?? repoId;
      if (!localPath) continue;
      try {
        const git = simpleGit(localPath);
        const diff = await git.diff(["HEAD"]);
        if (diff.trim().length > 0) {
          diffArtifacts.push({
            type: "diff",
            title: `Git diff — ${repoName}`,
            content: diff,
            content_type: "text/x-diff",
            sort_order: diffIdx++,
          });
        }
      } catch {
        // non-fatal
      }
    }

    const newStatus = result.success ? "done" : "failed";
    await dataClient.updateTask(task.id, {
      status: newStatus,
      result: result.output || result.error,
      started_at: startedAt,
      completed_at: completedAt,
    });

    // Persist all artifacts (diffs + agent-produced)
    const agentArtifacts = (result.artifacts ?? []).map((a, i) => ({
      ...a,
      sort_order: diffArtifacts.length + i,
    }));
    const allArtifacts = [...diffArtifacts, ...agentArtifacts];
    if (allArtifacts.length > 0) {
      try {
        await dataClient.createTaskArtifacts(task.id, allArtifacts);
      } catch (err) {
        console.error(`[Scheduler] createTaskArtifacts failed for task ${task.id}:`, err);
      }
    }

    this.emit(projectId, {
      type: "task_status",
      task_id: task.id,
      status: newStatus,
      agent_profile_id: task.agent_profile_id,
    });

    if (result.success) {
      this.emit(projectId, {
        type: "agent_completed",
        task_id: task.id,
        agent_role: task.type,
      });
    } else {
      this.emit(projectId, {
        type: "agent_failed",
        task_id: task.id,
        agent_role: task.type,
        error: result.error,
      });
    }

    // Push any commits the agent made — best-effort, non-fatal
    if (result.success) {
      await this.pushWorkspaces(projectId, task, workspaceMap, repositories);
    }

    inFlightIds.delete(task.id);
    terminalIds.add(task.id);

    if (result.success) {
      completedIds.add(task.id);
    } else {
      // Cascade-skip all transitive dependents of the failed task.
      // 1. Compute dependents locally (avoids extra DB round-trips for in-memory tracking).
      // 2. Persist to DB via the internal endpoint.
      // 3. Add to terminalIds and emit SSE so the loop and the frontend stay in sync.
      const dependents = this.getTransitiveDependents(task.id, allTasks, terminalIds);
      if (dependents.length > 0) {
        try {
          await dataClient.skipDependentTasks(task.id);
        } catch (err) {
          console.error(`[Scheduler] skipDependentTasks failed for task ${task.id}:`, err);
        }
        for (const depId of dependents) {
          terminalIds.add(depId);
          this.emit(projectId, {
            type: "task_status",
            task_id: depId,
            status: "skipped",
          });
        }
      }
    }

    void redis.publish(
      `project:${projectId}:task`,
      JSON.stringify({ task_id: task.id, status: newStatus })
    );

    OrchestrationService.getInstance().notifyTaskComplete(planId, task.id, result.success);
  }

  private emit(projectId: string, event: object): void {
    sseManager.broadcast(projectId, event);
    void redis.publish(`project:${projectId}:events`, JSON.stringify(event));
  }
}

File: agent-service-py/src/agent_service/agents/coordinator/scheduler.py
from __future__ import annotations

import asyncio
import json
import logging
import os
from typing import Dict, List, Optional, Set
from urllib.parse import urlparse

from agent_service.agents.coordinator.drivers.base import AgentArtifact, AgentTask
from agent_service.agents.coordinator.pool import AgentPool
from agent_service.config import config
from agent_service.core.redis import get_redis
from agent_service.crypto import decrypt
from agent_service.services import data_client, sse_manager
from agent_service.services.orchestration_service import OrchestrationService

logger = logging.getLogger(__name__)

TERMINAL_STATUSES = frozenset({"done", "failed", "skipped", "cancelled"})


class Scheduler:
    """
    Topological task scheduler. Reads tasks for a plan, resolves dependencies,
    and dispatches them to agent drivers respecting ``MAX_CONCURRENT_TASKS``.
    """

    def __init__(self, pool: AgentPool) -> None:
        self._pool = pool

    async def run(self, project_id: str, plan_id: str) -> None:
        await data_client.update_plan(plan_id, {"status": "executing"})
        self._emit(project_id, {"type": "plan_executing", "plan_id": plan_id})

        try:
            await self._run_internal(project_id, plan_id)
            await data_client.complete_plan_execution(plan_id)
            self._emit(project_id, {"type": "plan_completed", "plan_id": plan_id})
        except Exception as err:
            reason = str(err)
            try:
                await data_client.fail_plan_execution(plan_id, reason)
            except Exception:
                pass
            self._emit(project_id, {"type": "plan_failed", "plan_id": plan_id, "error": reason})
            raise

    async def _run_internal(self, project_id: str, plan_id: str) -> None:
        repositories = await data_client.get_project_repositories(project_id)
        tasks = await data_client.get_plan_tasks(plan_id)

        workspace_map = await self._clone_repositories(project_id, repositories)
        repo_by_id = {r["id"]: r for r in repositories}

        terminal_ids: Set[str] = set()
        completed_ids: Set[str] = set()
        in_flight_ids: Set[str] = set()

        for t in tasks:
            if t["status"] in TERMINAL_STATUSES:
                terminal_ids.add(t["id"])
            if t["status"] == "done":
                completed_ids.add(t["id"])

        max_concurrent = config.MAX_CONCURRENT_TASKS

        while len(terminal_ids) < len(tasks):
            ready = [
                t for t in tasks
                if t["id"] not in terminal_ids
                and t["id"] not in in_flight_ids
                and all(dep in completed_ids for dep in t.get("depends_on_task_ids", []))
            ]

            if not ready:
                if not in_flight_ids:
                    break  # deadlock / all done
                await asyncio.sleep(1.0)
                continue

            slots = max(0, max_concurrent - len(in_flight_ids))
            if slots == 0:
                await asyncio.sleep(1.0)
                continue

            batch = ready[:slots]
            await asyncio.gather(
                *[
                    self._dispatch_task(
                        project_id, plan_id, task, tasks,
                        workspace_map, repo_by_id,
                        terminal_ids, completed_ids, in_flight_ids,
                    )
                    for task in batch
                ],
                return_exceptions=True,
            )

    # ── Repository management ─────────────────────────────────────────────────

    def _build_clone_url(self, repo: dict) -> str:
        if repo.get("auth_type") == "token" and repo.get("credentials"):
            token = decrypt(repo["credentials"])
            parsed = urlparse(repo["remote_url"])
            host = parsed.hostname + (f":{parsed.port}" if parsed.port else "")
            return parsed._replace(netloc=f"{token}@{host}").geturl()
        return repo["remote_url"]

    async def _clone_repositories(
        self, project_id: str, repositories: List[dict]
    ) -> Dict[str, str]:
        workspace_map: Dict[str, str] = {}

        for repo in repositories:
            local_path = os.path.join(config.WORKSPACES_ROOT, project_id, repo["name"])
            clone_url = self._build_clone_url(repo)

            self._emit(project_id, {
                "type": "repo_status",
                "repo_id": repo["id"],
                "repo_name": repo["name"],
                "status": "cloning",
            })

            try:
                proc = await asyncio.create_subprocess_exec(
                    "git", "clone", clone_url, local_path,
                    "--branch", repo.get("branch") or "main",
                    stdout=asyncio.subprocess.DEVNULL,
                    stderr=asyncio.subprocess.PIPE,
                )
                _, stderr = await proc.communicate()
                if proc.returncode != 0:
                    # Repo already cloned — pull
                    pull = await asyncio.create_subprocess_exec(
                        "git", "-C", local_path, "pull",
                        stdout=asyncio.subprocess.DEVNULL,
                        stderr=asyncio.subprocess.DEVNULL,
                    )
                    await pull.wait()

                await data_client.update_repository_status(
                    repo["id"], {"status": "ready", "local_path": local_path}
                )
                workspace_map[repo["id"]] = local_path

                self._emit(project_id, {
                    "type": "repo_status",
                    "repo_id": repo["id"],
                    "repo_name": repo["name"],
                    "status": "ready",
                })
            except Exception as err:
                await data_client.update_repository_status(
                    repo["id"], {"status": "error", "error_message": str(err)}
                )
                self._emit(project_id, {
                    "type": "repo_status",
                    "repo_id": repo["id"],
                    "repo_name": repo["name"],
                    "status": "error",
                    "message": str(err),
                })

        return workspace_map

    # ── Git push ──────────────────────────────────────────────────────────────

    async def _push_workspaces(
        self, project_id: str, task: dict, workspace_map: Dict[str, str], repo_by_id: Dict[str, dict]
    ) -> None:
        for repo_id in task.get("repository_ids", []):
            local_path = workspace_map.get(repo_id)
            repo = repo_by_id.get(repo_id)
            if not local_path or not repo:
                continue
            try:
                # Stage and commit any uncommitted changes
                status_proc = await asyncio.create_subprocess_exec(
                    "git", "-C", local_path, "status", "--porcelain",
                    stdout=asyncio.subprocess.PIPE,
                    stderr=asyncio.subprocess.DEVNULL,
                )
                stdout, _ = await status_proc.communicate()
                if stdout.strip():
                    for cmd in [
                        ["git", "-C", local_path, "config", "user.email", "agent@swe-ai.local"],
                        ["git", "-C", local_path, "config", "user.name", "SWE AI Agent"],
                        ["git", "-C", local_path, "add", "."],
                        ["git", "-C", local_path, "commit", "-m", f'chore: agent result for task "{task["title"]}"'],
                    ]:
                        p = await asyncio.create_subprocess_exec(*cmd, stdout=asyncio.subprocess.DEVNULL, stderr=asyncio.subprocess.DEVNULL)
                        await p.wait()

                push_url = self._build_clone_url(repo)
                branch = repo.get("branch") or "main"
                push_proc = await asyncio.create_subprocess_exec(
                    "git", "-C", local_path, "push", push_url, branch,
                    stdout=asyncio.subprocess.DEVNULL,
                    stderr=asyncio.subprocess.DEVNULL,
                )
                await push_proc.wait()

                self._emit(project_id, {
                    "type": "repo_status",
                    "repo_id": repo["id"],
                    "repo_name": repo["name"],
                    "status": "ready",
                    "message": "Changes pushed to remote",
                })
            except Exception as err:
                logger.error("[Scheduler] Failed to push repo %s: %s", repo["name"], err)
                self._emit(project_id, {
                    "type": "repo_status",
                    "repo_id": repo["id"],
                    "repo_name": repo["name"],
                    "status": "ready",
                    "message": f"Push failed (non-fatal): {err}",
                })

    # ── Task dispatch ─────────────────────────────────────────────────────────

    async def _dispatch_task(
        self,
        project_id: str,
        plan_id: str,
        task: dict,
        all_tasks: List[dict],
        workspace_map: Dict[str, str],
        repo_by_id: Dict[str, dict],
        terminal_ids: Set[str],
        completed_ids: Set[str],
        in_flight_ids: Set[str],
    ) -> None:
        in_flight_ids.add(task["id"])
        try:
            await self._dispatch_task_inner(
                project_id, plan_id, task, all_tasks,
                workspace_map, repo_by_id,
                terminal_ids, completed_ids, in_flight_ids,
            )
        except Exception as err:
            logger.error("[Scheduler] Unexpected error in dispatch for task %s: %s", task["id"], err)
            in_flight_ids.discard(task["id"])
            terminal_ids.add(task["id"])
            try:
                await data_client.update_task(task["id"], {
                    "status": "failed",
                    "result": f"Scheduler internal error: {err}",
                    "completed_at": _now(),
                })
            except Exception:
                pass
            self._emit(project_id, {"type": "task_status", "task_id": task["id"], "status": "failed"})

    async def _dispatch_task_inner(
        self,
        project_id: str,
        plan_id: str,
        task: dict,
        all_tasks: List[dict],
        workspace_map: Dict[str, str],
        repo_by_id: Dict[str, dict],
        terminal_ids: Set[str],
        completed_ids: Set[str],
        in_flight_ids: Set[str],
    ) -> None:
        await data_client.update_task(task["id"], {"status": "in_progress"})
        self._emit(project_id, {
            "type": "task_status",
            "task_id": task["id"],
            "status": "in_progress",
            "agent_profile_id": task.get("agent_profile_id"),
        })

        workspaces: Dict[str, str] = {}
        for repo_id in task.get("repository_ids", []):
            local = workspace_map.get(repo_id)
            name = repo_by_id.get(repo_id, {}).get("name", repo_id)
            if local:
                workspaces[name] = local
        if not workspaces and workspace_map:
            workspaces["default"] = next(iter(workspace_map.values()))

        agent_task = AgentTask(
            id=task["id"],
            title=task["title"],
            description=task["description"],
            type=task["type"],
            agent_profile_id=task.get("agent_profile_id"),
        )

        driver = (
            self._pool.get_driver_by_role(task["type"])
            or (self._pool.get_driver(task["agent_profile_id"]) if task.get("agent_profile_id") else None)
        )

        self._emit(project_id, {
            "type": "agent_started",
            "task_id": task["id"],
            "agent_role": task["type"],
            "agent_profile_id": task.get("agent_profile_id"),
        })

        started_at = _now()

        if not driver:
            logger.warning("[Scheduler] No driver found for task %s type=%s", task["id"], task["type"])
            await data_client.update_task(task["id"], {
                "status": "failed",
                "result": "No driver found for task type or profile",
                "started_at": started_at,
                "completed_at": _now(),
            })
            self._emit(project_id, {"type": "task_status", "task_id": task["id"], "status": "failed"})
            in_flight_ids.discard(task["id"])
            terminal_ids.add(task["id"])
            return

        logger.info("[Scheduler] Executing task %s (%s) with driver", task["id"], task["title"])
        result = await driver.execute(agent_task, workspaces)
        logger.info("[Scheduler] Task %s finished: success=%s", task["id"], result.success)

        completed_at = _now()

        # Capture git diff
        diff_artifacts: list = []
        diff_idx = 0
        for repo_id in task.get("repository_ids", []):
            local = workspace_map.get(repo_id)
            repo_name = repo_by_id.get(repo_id, {}).get("name", repo_id)
            if not local:
                continue
            try:
                proc = await asyncio.create_subprocess_exec(
                    "git", "-C", local, "diff", "HEAD",
                    stdout=asyncio.subprocess.PIPE,
                    stderr=asyncio.subprocess.DEVNULL,
                )
                stdout, _ = await proc.communicate()
                diff = stdout.decode(errors="replace").strip()
                if diff:
                    diff_artifacts.append({
                        "type": "diff",
                        "title": f"Git diff — {repo_name}",
                        "content": diff,
                        "content_type": "text/x-diff",
                        "sort_order": diff_idx,
                    })
                    diff_idx += 1
            except Exception:
                pass

        new_status = "done" if result.success else "failed"
        await data_client.update_task(task["id"], {
            "status": new_status,
            "result": result.output or result.error,
            "started_at": started_at,
            "completed_at": completed_at,
        })

        all_artifacts = diff_artifacts + [
            {
                "type": a.type,
                "title": a.title,
                "content": a.content,
                "content_type": a.content_type,
                "metadata": a.metadata,
                "sort_order": len(diff_artifacts) + i,
            }
            for i, a in enumerate(result.artifacts or [])
        ]
        if all_artifacts:
            try:
                await data_client.create_task_artifacts(task["id"], all_artifacts)
            except Exception as err:
                logger.error("[Scheduler] create_task_artifacts failed for task %s: %s", task["id"], err)

        self._emit(project_id, {
            "type": "task_status",
            "task_id": task["id"],
            "status": new_status,
            "agent_profile_id": task.get("agent_profile_id"),
        })

        if result.success:
            self._emit(project_id, {"type": "agent_completed", "task_id": task["id"], "agent_role": task["type"]})
            await self._push_workspaces(project_id, task, workspace_map, repo_by_id)
        else:
            self._emit(project_id, {
                "type": "agent_failed",
                "task_id": task["id"],
                "agent_role": task["type"],
                "error": result.error,
            })

        in_flight_ids.discard(task["id"])
        terminal_ids.add(task["id"])

        if result.success:
            completed_ids.add(task["id"])
        else:
            dependents = self._get_transitive_dependents(task["id"], all_tasks, terminal_ids)
            if dependents:
                try:
                    await data_client.skip_dependent_tasks(task["id"])
                except Exception as err:
                    logger.error("[Scheduler] skip_dependent_tasks failed: %s", err)
                for dep_id in dependents:
                    terminal_ids.add(dep_id)
                    self._emit(project_id, {"type": "task_status", "task_id": dep_id, "status": "skipped"})

        redis = get_redis()
        await redis.publish(
            f"project:{project_id}:task",
            json.dumps({"task_id": task["id"], "status": new_status}),
        )

        OrchestrationService.get_instance().notify_task_complete(plan_id, task["id"], result.success)

    def _get_transitive_dependents(
        self, task_id: str, tasks: List[dict], exclude_ids: Set[str]
    ) -> List[str]:
        result: list[str] = []
        visited: Set[str] = set()

        def traverse(tid: str) -> None:
            direct = [t for t in tasks if tid in t.get("depends_on_task_ids", []) and t["id"] not in exclude_ids]
            for dep in direct:
                if dep["id"] not in visited:
                    visited.add(dep["id"])
                    result.append(dep["id"])
                    traverse(dep["id"])

        traverse(task_id)
        return result

    def _emit(self, project_id: str, event: dict) -> None:
        sse_manager.broadcast(project_id, event)
        redis = get_redis()
        asyncio.ensure_future(
            redis.publish(f"project:{project_id}:events", json.dumps(event))
        )


def _now() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).isoformat()
